lib/SILOptimizer/Transforms/DeadCodeElimination.cpp

```diff
@@ -306,9 +306,8 @@ void DCE::markLive() {
             disableBorrowDCE(root);
           }
         }
-        // If we have a lexical borrow scope or a pointer escape, disable DCE.
-        if (borrowInst->isLexical() ||
-            hasPointerEscape(BorrowedValue(borrowInst))) {
+        // If we have a pointer escape, disable DCE.
+        if (hasPointerEscape(BorrowedValue(borrowInst))) {
           disableBorrowDCE(borrowInst);
```

Comment:
DCE of lexical borrows were turned off conservatively. I don't see a reason we can't shorten a lexical borrows lifetime via a dead phi. Let me know if you think otherwise @nate-chandler @atrick

Reply:
Can eliminating such a borrow result in allowing a deinit to be hoisted over a deinit barrier?  For example, in the test case if the last block looked as follows: 

```
 bb3(%11 : @owned $Klass, %12 : @guaranteed $Klass, %13 : @owned $Klass):
   destroy_value %13 : $Klass
   some_deinit_barrier
   end_borrow %12 : $Klass
   destroy_value %11 : $Klass
   %17 = tuple ()
   return %17 : $()
 }
```

Reply:
I think we can only do this if the borrow introducer is also lexical. But @nate-chandler has an optimization to remove the lexical flag from the inner scope in those cases. I might not work yet for phis/reborrows. To do this right I think we need to start using the visitBorrowIntroducers/visitEnclosingDef/visitOutermostEnclosingDef set of utilities